### PR TITLE
[HW][IRN/IST] Make hw modules IST's, add verif pass.

### PR DESCRIPTION
**Pull request by dtzSiFive**
Until have proper container op to put the trait on, add a dummy struct we can classof to classify IRN-like, and add a simple pass that drives verification on these if they aren't self-verifying.

Add verification pass to pipeline, basic test for duplicate inner symbols and invalid InnerRef user.

cc #3526 (nearly fixes), #4453.
cc #5716 .

### CODE REVIEW

File: tools/firtool/firtool.cpp [hunk 1]
```diff
@@ -381,6 +382,9 @@ static LogicalResult processBuffer(
     if (exportModuleHierarchy)
       pm.addPass(sv::createHWExportModuleHierarchyPass(outputFilename));
 
+    // Check inner symbols and inner refs.
+    pm.addPass(hw::createVerifyInnerRefNamespacePass());
```
Comment (seldridge) [inline tools/firtool/firtool.cpp:386]: The idea here is that this will catch some errors that could manifest during `ExportVerilog`? This may miss some issues. It is better than nothing.

I'm lukewarm on this approach as it reminds me of SFC-style verification. We used to do all the verification in a [block of serial passes](https://github.com/chipsalliance/firrtl/blob/1.6.x/src/main/scala/firrtl/stage/Forms.scala#L33) that was run once and never again. I ran experiments where these passes would be rescheduled after every pass (normal "verification") and it uncovered a lot of really weird issues.

tl;dr: I'm okay with this, but only as an intermediate step. Don't let this hang around for too long. 😉 We need to bite the bullet and get the container op in place.

**Review by seldridge — APPROVED** [review]
I'm fine with this as an intermediate step. What's the lead time to get the container op in place?

**Review by fabianschuiki — APPROVED** [review]
I think this is a good stepping stone!

Just out of curiosity: have you considered making the verifier for IST/IS/IR walk up to the nearest `InnerRefNamespace` or `mlir::ModuleOp` and use `getAnalysis(...)` to get a symbol table? Does that even work from a verifier?

File: tools/firtool/firtool.cpp (hunk 1 above)
Comment (teqdruid) [reply tools/firtool/firtool.cpp:386]: +1 on the container op

File: include/circt/Dialect/HW/HWOpInterfaces.h [hunk 2]
```diff
@@ -222,6 +222,17 @@ namespace detail {
 LogicalResult verifyInnerRefNamespace(Operation *op);
 } // namespace detail
 
+/// Classify operations that are InnerRefNamespace-like,
+/// until structure is in place to do this via Traits.
+/// Useful for getParentOfType<>, or scheudling passes.
```
Comment (rwy7) [inline include/circt/Dialect/HW/HWOpInterfaces.h:227]:
```suggestion
/// Useful for getParentOfType<>, or scheduling passes.
```

**Review by rwy7 — APPROVED** [review]

File: tools/firtool/firtool.cpp (hunk 1 above)
Comment (dtzSiFive) [reply tools/firtool/firtool.cpp:386]: 100% I'd prefer the container op.  However we're increasingly using this and getting some verification in place seems high-priority to avoid things being too broken without noticing.

> The idea here is that this will catch some errors that could manifest during ExportVerilog? This may miss some issues. It is better than nothing.

Yes, idea was to schedule it in a few places to help catch things -- right now this PR adds it as part of the FIRRTL->HW pipeline, the HW->SV pipeline, and yep a final run before `ExportVerilog`.

Probably could automagically insert this between all passes or something (drive via pass instrumentation?) if the concern is only having a one-off check being performed.  Strong agree I'd prefer to verify all the things regularly :+1:.

As an aside there's a lot of "walk this module looking for the symbol" logic that can be improved once all this in place.

> tl;dr: I'm okay with this, but only as an intermediate step. Don't let this hang around for too long. 😉 We need to bite the bullet and get the container op in place.

Absolutely and thanks for saying so! The op itself is easy, getting everything to use and maintain that sense of structure vs bag of operations is a bit more challenging.  Haven't pushed on it for a few days due to other demands, but will revisit and see if can't form a migration plan. :+1: .

**Comment by dtzSiFive** [comment]
> Just out of curiosity: have you considered making the verifier for IST/IS/IR walk up to the nearest `InnerRefNamespace` or `mlir::ModuleOp` and use `getAnalysis(...)` to get a symbol table? Does that even work from a verifier?

Can't use `getAnalysis` from verifiers presently, upstream's stance is that verifiers shouldn't be needing anything heavy enough to warrant an analysis IIRC, esp a cached one (perhaps verification shouldn't assume passes accurately preserve the analyses they claim to?).

Similarly I'm not sure that an analysis can reasonably report diagnostics/fail, but haven't pushed on that.

The problem is walking the entire design to gather all inner symbol definitions is expensive and we only want to do it once-- hence the top-down verification strategy where a single place builds up the information and feeds to to all ops implementing the verification hook.

The little struct w/classof is meant to be a placeholder for the container op re:"walk up to the IRN or ModuleOp" -- and something like it will likely be useful to stay flexible to whatever container op is being used, especially as the container op lands and we migrate over (moving at all at once is a lot).  We'll see :+1: .

File: include/circt/Dialect/HW/HWOpInterfaces.h (hunk 2 above)
Comment (dtzSiFive) [reply include/circt/Dialect/HW/HWOpInterfaces.h]: Thanks!
